app/models/endpoint.rb

```diff
@@ -16,7 +16,31 @@ class Endpoint < ApplicationRecord
   validates :endpoint_type, presence: true
   validates :endpoint_node, presence: true
   validates :storage_location, presence: true
-  validates :recovery_cost, presence: true
+
+  scope :archive, lambda {
+    # TODO: maybe endpoint_class should be an enum or a constant?
+    joins(:endpoint_type).where(endpoint_types: { endpoint_class: 'archive' })
+  }
+
+  # for the given druid, which archive endpoints should have preserved copies?
+  scope :archive_targets, lambda { |druid|
+    archive.joins(preservation_policies: [:preserved_objects]).where(preserved_objects: { druid: druid })
+  }
+
+  scope :which_need_archive_copy, lambda { |druid, version|
+    # cast version to int for nicer errors in the case of bad input.  though ActiveRecord/ARel would still protect
+    # against injection attacks even without using a bind var (e.g. a string passed in for an int col query would
+    # silently be inserted as 0, an INTEGER).
+    pc_table = PreservedCopy.arel_table
+    ep_table = Endpoint.arel_table
+    endpoint_has_pres_copy_subquery =
+      PreservedCopy.where(
+        pc_table[:endpoint_id].eq(ep_table[:id])
+          .and(pc_table[:version].eq(version.to_i))
+      ).exists
```

Comment:
another way to write this subquery would be:
```ruby
  pc_table = PreservedCopy.arel_table
  ep_table = Endpoint.arel_table
  endpoint_has_pres_copy_subquery =
    PreservedCopy.where(
      'preserved_copies.endpoint_id = endpoints.id and preserved_copies.version = #{version.to_i}'
    ).exists
```

from the reading i found online (stack exchange posts and blog posts), the `arel_table` approach in this PR seemed slightly more idiomatic, and *maybe* more maintainable in theory, but it might also be slightly less readable?

do others have strong preferences?

Reply:
In either case, the subquery only makes sense in the context of the macro-query `.joins`.  I can take a stab at some other forms if you want.

Reply:
I don't have an opinion on which way to go with this.  Are we likely to care about performance for this?  I think the two methods John proposes are equally readable/maintainable.

Reply:
Query structure ends up looking like:
```
SELECT "endpoints".* FROM "endpoints" 
INNER JOIN "endpoint_types" ON "endpoint_types"."id" = "endpoints"."endpoint_type_id" 
INNER JOIN "endpoints_preservation_policies" ON "endpoints_preservation_policies"."endpoint_id" = "endpoints"."id" 
INNER JOIN "preservation_policies" ON "preservation_policies"."id" = "endpoints_preservation_policies"."preservation_policy_id" 
INNER JOIN "preserved_objects" ON "preserved_objects"."preservation_policy_id" = "preservation_policies"."id" 
WHERE "endpoint_types"."endpoint_class" = $1 
AND "preserved_objects"."druid" = $2 
AND (NOT (EXISTS (
  SELECT "preserved_copies".* FROM "preserved_copies" 
  WHERE ("preserved_copies"."endpoint_id" = "endpoints"."id" 
  AND "preserved_copies"."version" = 1)
)))
```

Example [`EXPLAIN ANALYZE`](https://explain.depesz.com/s/sAH8) for my tiny dev db.

Reply:
yup, that looks like the SQL i remember seeing get generated.  and that plan looks almost exactly like the one in my notes, with very slightly different overall costs, but the same join etc plan (which makes sense, since we have the same fixture data on similar laptops).